design/src/rfcs/rfc0022_error_context_and_compatibility.md

```diff
@@ -0,0 +1,402 @@
+    },
+
+    // Bad: The error source type is public, and even though its a boxed error, it won't
+    // be possible to change it to an opaque error type later (for example, if/when
+    // opaque errors become practical due to standard library stabilizations).
+    #[non_exhaustive]
+    VariantG {
+        source: Box<dyn Error + Send + Sync + 'static>,
+    }
+}
+
+pub struct VariantA {
+    some_field: u32,
+    // This is private, so it's fine to reference the external library's error type
+    source: http::uri::InvalidUri
+}
+
+impl VariantA {
+    fn some_field(&self) -> u32 {
+        self.some_field
+    }
+}
+```
+
+Error variants that contain a source must return it from the `Error::source` method.
+The `source` implementation _should not_ use the catch all (`_`) match arm, as this makes it easy to miss
+adding a new error variant's source at a later date.
+
+The error `Display` implementation _must not_ include the source in its output:
+
+```rust
+// Good
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::VariantA => write!(f, "variant a"),
+            Self::VariantB { some_additional_info, .. } => write!(f, "variant b ({some_additional_info})"),
+            // ... and so on
+        }
+    }
+}
+
+// Bad
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::VariantA => write!(f, "variant a"),
+            // Bad: includes the source in the `Display` output, which leads to duplicate error information
+            Self::VariantB { some_additional_info, source } => write!(f, "variant b ({some_additional_info}): {source}"),
+            // ... and so on
+        }
+    }
+}
+```
+
+### Informative error pattern
+
+Informative errors must be represented as structs. If error messaging changes based on an underlying cause, then a
+private error kind enum can be used internally for this purpose. For example:
+
+```rust
+#[derive(Debug)]
+pub struct InformativeError {
+    some_additional_info: u32,
+    source: AnotherError,
+}
+
+impl fmt::Display for InformativeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "some informative message with {}", self.some_additional_info)
+    }
+}
+
+impl Error for InformativeError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(&self.source)
+    }
+}
+```
+
+In general, informative errors should be referenced by variants in actionable errors since they cannot be converted
+to actionable errors at a later date without a breaking change. This is not a hard rule, however. Use your best judgement
+for the situation.
+
+### Displaying full error context
+
+In code where errors are logged rather than returned to the customer, the full error source chain
+must be displayed. This will be made easy by placing a `DisplayErrorContext` struct in `aws-smithy-types` that
```

Comment:
Hadn't considered including that in `DisplayErrorContext`, but it seems like a good idea (it could follow the full error message in parenthesis or something).